### PR TITLE
87:  PUT /group/{id} endpoint and basic authorization

### LINKED ISSUE #87: Add PUT /api/group/{id}

### PR DESCRIPTION
**Pull request by GR0ZA**
I've added a PUT endpoint for groups. Nothing spectacular. For convenience only name and the owner of the group can be changed.

Much more interesting point is the authorization approach implemented for this endpoint for the first time in our app. 

It's pretty simple:
- our `Authenticate` middleware in `cookies.go` stores the userID retrieved from the cookie in the context (`c.Locals("userID", sessionCookie.UserID`)
- we access this `userID` in `group_handler` and pass it to the service
- service retrieves the group that has to be updated and checks whether the group's owner is equal to `userID` from the cookie. If not a "not authorized" error message is returned.

The assumption is that only the group owner can update it.

You may wonder: why do we pass an extra parameter to the service instead of authorizing directly in the handler? The reason is: handlers should be dumb as fuck. For the authorization we need to retrieve the existing group first (to compare its owner with the requester ID). We don't want a handler to do this. He should only pass needed things to the service. Service is where the magic happens.
 

Closes #87 and deals with #48.

### CODE REVIEW

File: presentation/handler/group_handler.go [hunk 1]
```diff
@@ -52,6 +52,46 @@ func (h GroupHandler) Create(c *fiber.Ctx) error {
...
+	uid, err := uuid.Parse(id)
+	if err != nil {
+		return core.Error(c, fiber.StatusBadRequest, fmt.Sprintf(ErrMsgParseUUID, uid, err))
+	}
+
+	// parse request
+	var request GroupInputDTO
+	if err := c.BodyParser(&request); err != nil {
+		return core.Error(c, fiber.StatusBadRequest, fmt.Sprintf(ErrMsgGroupParse, err))
+	}
```
Comment (f-odc) [inline presentation/handler/group_handler.go:82]: Why do we use GroupInputDTO for the Update method. For Update Group we should be able to update all group fields not only the owner ID and the name.

File: authentication/cookies.go [hunk 2]
```diff
@@ -53,6 +53,9 @@ func (a *Authenticator) Authenticate(c *fiber.Ctx) error {
 		return core.Error(c, fiber.StatusUnauthorized, fmt.Sprintf(ErrMsgAuthentication, err))
 	}
 
+	// set userID in context
+	c.Locals("userID", sessionCookie.UserID)
```
Comment (f-odc) [inline authentication/cookies.go:57]: Let us use a variable for userID here.

File: domain/service/impl/errors.go
Comment (f-odc) [inline domain/service/impl/errors.go:1]: I am not sure if we should create a own file for the service errors. Especially if we only have one error.

We have connection.go for storage errors, we have messages.go for handler errors and now errors.go for service errors. Please let us make this uniform.

File: presentation/handler/group_handler.go [hunk 3]
```diff
@@ -52,6 +52,46 @@ func (h GroupHandler) Create(c *fiber.Ctx) error {
...
+
+	userID := c.Locals("userID").(uuid.UUID)
```
Comment (f-odc) [inline presentation/handler/group_handler.go:84]: Here use also the stored variable for "userID".

**Review by f-odc — CHANGES_REQUESTED** [review]

File: presentation/handler/group_handler.go (hunk 1 above)
Comment (GR0ZA) [reply presentation/handler/group_handler.go:82]: It's the same case as in `AddItem` and `ChangeItem`. There we send a `ItemInputDTO`, both for creating and updating an item. This is a common scenario that DTOs for creating and updating a resource are very similar (or equal). Using `GroupInputDTO` is not wrong here.

Group fields are:
- `ID` -> not contained in body
- `Name`  -> sent on POST and PUT
- `Owner`  -> sent on POST and PUT 
- `Bills`  -> not contained in body (add, update & remove bills handled via /bill endpoint)
- `Members`  -> missing on both POST and PUT

So the only field missing for an update is `Members` . We have 2 options: have two distinct DTOs (one for update, one for create) or one DTO with the missing Members. In general one DTO sounds better. A user can join a group only after accepting the invitation, so it's some extra logic we have to keep in mind. I don't want to deal with it here, I'll create a new issue for that.

Comment (GR0ZA) [reply presentation/handler/group_handler.go:83]: #108

File: domain/service/impl/errors.go
Comment (GR0ZA) [reply domain/service/impl/errors.go:1]: As discussed removed Connection interface and renamed connection.go to errors.go

**Pull request merged**